pulsar-common/src/main/java/org/apache/pulsar/common/util/collections/ConcurrentRoaringBitSet.java

```diff
@@ -0,0 +1,459 @@
+            // Fallback to read lock
+            stamp = rwLock.readLock();
+            try {
+                byteArray = super.toByteArray();
+            } finally {
+                rwLock.unlockRead(stamp);
+            }
+        }
+        return byteArray;
+    }
+
+    @Override
+    public long[] toLongArray() {
+        long stamp = rwLock.tryOptimisticRead();
+        long[] longArray = super.toLongArray();
+        if (!rwLock.validate(stamp)) {
+            // Fallback to read lock
+            stamp = rwLock.readLock();
+            try {
+                longArray = super.toLongArray();
+            } finally {
+                rwLock.unlockRead(stamp);
+            }
+        }
+        return longArray;
+    }
+
+    @Override
+    public void flip(int bitIndex) {
+        long stamp = rwLock.writeLock();
+        try {
+            super.flip(bitIndex);
+        } finally {
+            rwLock.unlockWrite(stamp);
+        }
+    }
+
+    @Override
+    public void flip(int fromIndex, int toIndex) {
+        long stamp = rwLock.writeLock();
+        try {
+            super.flip(fromIndex, toIndex);
+        } finally {
+            rwLock.unlockWrite(stamp);
+        }
+    }
+
+    @Override
+    public void set(int bitIndex, boolean value) {
+        long stamp = rwLock.writeLock();
+        try {
+            super.set(bitIndex, value);
+        } finally {
+            rwLock.unlockWrite(stamp);
+        }
+    }
+
+    @Override
+    public void set(int fromIndex, int toIndex, boolean value) {
+        long stamp = rwLock.writeLock();
+        try {
+            super.set(fromIndex, toIndex, value);
+        } finally {
+            rwLock.unlockWrite(stamp);
+        }
+    }
+
+    @Override
+    public RoaringBitSet get(int fromIndex, int toIndex) {
+        long stamp = rwLock.tryOptimisticRead();
+        RoaringBitSet bitSet = super.get(fromIndex, toIndex);
+        if (!rwLock.validate(stamp)) {
+            // Fallback to read lock
+            stamp = rwLock.readLock();
+            try {
+                bitSet = super.get(fromIndex, toIndex);
+            } finally {
+                rwLock.unlockRead(stamp);
+            }
+        }
+        return bitSet;
+    }
+
+    /**
+     * Thread-safe version of {@code length()}.
+     * StampedLock is not reentrant and that's why the length() method is not overridden. Overriding length() method
+     * would require to use a reentrant lock which would be less performant.
+     *
+     * @return length of the bit set
+     */
+    public int safeLength() {
+        long stamp = rwLock.tryOptimisticRead();
+        int length = super.length();
+        if (!rwLock.validate(stamp)) {
+            // Fallback to read lock
+            stamp = rwLock.readLock();
+            try {
+                length = super.length();
+            } finally {
+                rwLock.unlockRead(stamp);
+            }
+        }
+        return length;
+    }
```

Comment:
this doesn't seem to be needed for ConcurrentRoaringBitSet. This was added as a workaround for the re-entrant calls that BitSet implementation makes when calling `.length()`. `StampedLock` isn't a re-entrant lock.